src/CharRNN/index.js

```diff
@@ -0,0 +1,245 @@
+    if (!stateful) {
+      this.state = this.zeroState;
+    }
+
+    const results = [];
+    const userInput = Array.from(seed);
+    const encodedInput = [];
+
+    userInput.forEach((char) => {
+      encodedInput.push(this.vocab[char]);
+    });
+
+    let input = encodedInput[0];
+    let probabilitiesNormalized = []; // will contain final probabilities (normalized)
+
+    for (let i = 0; i < userInput.length + length + -1; i += 1) {
+      const onehotBuffer = tf.buffer([1, this.vocabSize]);
+      onehotBuffer.set(1.0, 0, input);
+      const onehot = onehotBuffer.toTensor();
+      let output;
+      if (this.model.embedding) {
+        const embedded = tf.matMul(onehot, this.model.embedding);
+        output = tf.multiRNNCell(this.cells, embedded, this.state.c, this.state.h);
+      } else {
+        output = tf.multiRNNCell(this.cells, onehot, this.state.c, this.state.h);
+      }
+
+      this.state.c = output[0];
+      this.state.h = output[1];
+
+      const outputH = this.state.h[1];
+      const weightedResult = tf.matMul(outputH, this.model.fullyConnectedWeights);
+      const logits = tf.add(weightedResult, this.model.fullyConnectedBiases);
+      const divided = tf.div(logits, tf.tensor(temperature));
+      const probabilities = tf.exp(divided);
+      probabilitiesNormalized = await tf.div(
+        probabilities,
+        tf.sum(probabilities),
+      ).data();
+
+      if (i < userInput.length - 1) {
+        input = encodedInput[i + 1];
+      } else {
+        input = sampleFromDistribution(probabilitiesNormalized);
+        results.push(input);
+      }
+    }
+
+    let generated = '';
+    results.forEach((char) => {
+      const mapped = Object.keys(this.vocab).find(key => this.vocab[key] === char);
+      if (mapped) {
+        generated += mapped;
+      }
+    });
+    this.probabilities = probabilitiesNormalized;
+    return {
+      sample: generated,
+      state: this.state,
+      probabilities: this.probabilities,
+    };
+  }
+
+  reset() {
+    this.state = this.zeroState;
+  }
+
+  // stateless
+  async generate(options, callback) {
+    this.reset();
+    return callCallback(this.generateInternal(options), callback);
+  }
+
+  // stateful
+  async predict(temperature, callback) {
+    let probabilitiesNormalized = [];
+    const outputH = this.state.h[1];
+    const weightedResult = tf.matMul(outputH, this.model.fullyConnectedWeights);
+    const logits = tf.add(weightedResult, this.model.fullyConnectedBiases);
+    const divided = tf.div(logits, tf.tensor(temperature));
+    const probabilities = tf.exp(divided);
+    probabilitiesNormalized = await tf.div(
+      probabilities,
+      tf.sum(probabilities),
+    ).data();
+
+    const sample = sampleFromDistribution(probabilitiesNormalized);
+    const result = Object.keys(this.vocab).find(key => this.vocab[key] === sample);
+    if (callback) {
+      callback(result);
+    }
+    return result;
```

Comment:
This is where I think we should make a nicely formatted object or array with a dictionary of chars -> probabilities!

Reply:
fixed!